### ISSUE TITLE
config error: dyninst 11.0.0 doesn't build on theta-gpu at ANL

### ISSUE THREAD

**Issue by mwkrentel** [dyninst/dyninst#1025]
I looked at the dyninst 11.0.0 build problem again that I mentioned on
Wednesday's call.  It's very mysterious, but reproducible.

I've only seen this happen on the theta-gpu compute nodes at Argonne.
First, you have to login to theta at ANL.  Then, login to theta-gpu
service node (thetagpusn1 or thetagpusn2).  That should be automatic
via rhost (no password).  Finally, get to theta-gpu compute node via
interactive qsub.

```
qsub -q full-node -n 1 -t 200 -A Comp_Perf_Workshop -I
```

You could also use '-q single-gpu' but it's limited to 60 minutes.
  
  https://www.alcf.anl.gov/support-center/theta-gpu-nodes/getting-started-thetagpu

Right now, you don't have an allocation for theta-gpu, but I'm asking
Ray Loy to get you one (and to renew mine beyond the workshop).

In brief, on theta-gpu, the build for dyninst fails, unable to find
libiberty.a.  (Yes, they tell you to build on theta-gpu compute
nodes.)

There's a spack module for gcc 10.2.0, I was using that.  I used the
latest libiberty, 2.33.1 (I should update that, but whatever).
Cmake in /usr/bin is 3.10.2, so I used that as a spack external.

It fails, but the error message just says, missing library.

```
-- Could NOT find LibIberty (missing: LibIberty_LIBRARIES) 
CMake Error at cmake/LibIberty.cmake:53 (message):
  LibIberty not found and cannot be downloaded because build is sterile.
Call Stack (most recent call first):
  CMakeLists.txt:34 (include)
```

Adding -DLibIberty_LIBRARIES=...../lib64/libiberty.a makes it work,
but this has never been necessary before.

Notes:

(1) I'm seriously at a loss to explain what's different here.  On no
other machine have I seen this problem.  Note: theta-gpu is not a
Cray/Prg-Env build.  Treat it like regular Linux.

The spack install does have ...../lib64/libiberty.a, the same as other
machines where it works, ie, in lib64.

(2) theta-gpu is running Ubuntu 18.04.  Is there anything in cmake
that would be specific to the Linux distro?  Something that might say
on this machine try lib64 but this other machine only try lib??

(3) It fails the same way, with the same message building directly
from cmake, outside of spack.

(4) It's not the version of cmake.  I've tried various cmake from 3.10
to 3.12, 3.15 to 3.19 and they all either work or all don't work,
depending on the machine.

(5) I've checked other machines where it works and theta-gpu.  Neither
of them have /usr/include/libiberty.h or /usr/lib64/libiberty.a, so I
don't think the problem is that it's actually picking up a system
libiberty and not telling anyone.

(6) If you want to compare this on a machine where it works, I'd
suggest NOT to use the theta front end.  It was getting mixed up with
Cray PrgEnv things and strange results like not finding TBB.

In short, the cmake recipes are not finding libiberty.a but in a way
that I don't understand and never seen before.

@hainest

**Comment by hainest** [comment]
Were you able to get in contact with Ray? I've tried logging into a theta-gpu machine a few times, but I've never been able to.

**Comment by mwkrentel** [comment]
@hainest I guess Ray didn't get back to you?
I ping'd him Monday, but I also haven't heard back.

**Comment by hainest** [comment]
@mwkrentel I've not heard from him, and I can't log into the theta-gpu partition.

**Comment by mwkrentel** [comment]
@hainest I finally heard from Ray Loy today, and we should now be able
to access the compute nodes via the Tools account.

**Comment by hainest** [comment]
@mwkrentel I've logged into a thetagpu login node, but I'm still setting up my environment there. They are evidently using Ubuntu on the gpu nodes, but SLES on the theta nodes. That's causing some troubles with my default environment configs. I'll work on it and get back to you. How were you building Dyninst on theta-gpu? Just through spack?

**Comment by mwkrentel** [comment]
> How were you building Dyninst on theta-gpu? Just through spack?

@hainest I believe I mentioned this above in (3).
It fails the same way either with spack or directly via cmake.

**Comment by mwkrentel** [comment]
@hainest I've gotten two more reports of this bug this week, so let's
prioritize this if possible.

**Comment by mwkrentel** [comment]
Well, this is bad.  Theta-gpu has been updated to Ubuntu 20.04.2 and
now it works for me.

Aaron Cherian in our group also reported the problem.  When he tells
me what machine, I'll test it there.

**Comment by mwkrentel** [comment]
@hainest Aaron's machine is at some company where he's working for the summer.
But it is running Ubuntu 18.04, same as the old theta-gpu.  The third
machine is in the cloud, but Debian 11, so close to Ubuntu in some respects.

Do you have access to Ubuntu 18.04?
It seems like that's where we should focus.

**Comment by hainest** [comment]
@mwkrentel I routinely test on an Ubuntu-18.04 x86_64 machine with no issues. It uses spack to manage Dyninst's dependencies, but builds Dyninst from source manually. I'll see if I can get a container install of Debian 11 to test there.

**Comment by hainest** [comment]
I just ran a build on my usual 18.04 test box. CMake is able to find the libiberty from my spack installation

```
-- Found LibIberty: /nobackup/thaines/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-8.4.0/libiberty-2.33.1-2tb5tfhvtmqswq4nfmnyhrasarhr36a7/lib/libiberty.a  
-- LibIberty include dirs: /nobackup/thaines/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-8.4.0/libiberty-2.33.1-2tb5tfhvtmqswq4nfmnyhrasarhr36a7/include/libiberty
-- LibIberty library dirs: /nobackup/thaines/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-8.4.0/libiberty-2.33.1-2tb5tfhvtmqswq4nfmnyhrasarhr36a7/lib
-- LibIberty libraries: /nobackup/thaines/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-8.4.0/libiberty-2.33.1-2tb5tfhvtmqswq4nfmnyhrasarhr36a7/lib/libiberty.a
```
and (after unloading the spack module), I can use the one from the system

```
-- Found LibIberty: /usr/lib/x86_64-linux-gnu/libiberty.a
-- LibIberty include dirs: /usr/include/libiberty
-- LibIberty library dirs: /usr/lib/x86_64-linux-gnu
-- LibIberty libraries: /usr/lib/x86_64-linux-gnu/libiberty.a
```

**Comment by hainest** [comment]
I tried this on the Docker image for Debian 11 without trouble (although the elfutils package doesn't install libelf.h, so I had to manually add it):

```
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

```
apt install build-essential gcc-10 g++-10 cmake git libboost-all-dev libiberty-dev libtbb-dev elfutils
```

Building Dyninst my usual way:

```
Found LibIberty: /usr/lib/x86_64-linux-gnu/libiberty.a  
-- LibIberty include dirs: /usr/include/libiberty
-- LibIberty library dirs: /usr/lib/x86_64-linux-gnu
-- LibIberty libraries: /usr/lib/x86_64-linux-gnu/libiberty.a
```

**Comment by mwkrentel** [comment]
Ah, but what happens in an all-spack build?

I'm developing a theory.  Is CONFIG_SITE set in the environment on the
Debian 11 machine?

Autotools responds to CONFIG_SITE and it's possible that some site has
CONFIG_SITE set to something that cmake doesn't understand.  Probably,
I could always set CONFIG_SITE to something ridiculous that would
break cmake.  But what I'm looking for is something that is not
unreasonable but something that cmake doesn't expect.

**Comment by hainest** [comment]
I'm not seeing that with Debian 11

```
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

# spack install dyninst
... // all the dependencies
==> Installing dyninst-11.0.1-2danmmfcntvv5kvpigdv45nxakh3tx76
==> No binary for dyninst-11.0.1-2danmmfcntvv5kvpigdv45nxakh3tx76 found: installing from source
==> No patches needed for dyninst
==> dyninst: Executing phase: 'cmake'
==> dyninst: Executing phase: 'build'
==> dyninst: Executing phase: 'install'
==> dyninst: Successfully installed dyninst-11.0.1-2danmmfcntvv5kvpigdv45nxakh3tx76
  Fetch: 5.23s.  Build: 7m 16.73s.  Total: 7m 21.96s.
[+] /spack/opt/spack/linux-debian11-piledriver/gcc-10.2.1/dyninst-11.0.1-2danmmfcntvv5kvpigdv45nxakh3tx76
```

and the same on Ubuntu 18.04

```
==> Installing dyninst-11.0.1-5oytk3axr35ohmlipkvnynb6cm3intg6
==> No binary for dyninst-11.0.1-5oytk3axr35ohmlipkvnynb6cm3intg6 found: installing from source
==> No patches needed for dyninst
==> dyninst: Executing phase: 'cmake'
==> dyninst: Executing phase: 'build'
==> dyninst: Executing phase: 'install'
==> dyninst: Successfully installed dyninst-11.0.1-5oytk3axr35ohmlipkvnynb6cm3intg6
  Fetch: 2.61s.  Build: 8m 38.13s.  Total: 8m 40.74s.
[+] /nobackup/thaines/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-10.2.0/dyninst-11.0.1-5oytk3axr35ohmlipkvnynb6cm3intg6
```

**Comment by mwkrentel** [comment]
Hmm, at this point, I don't have a good way to reproduce the problem.
Maybe Max Dechantsreiter could set us up with an identical cloud system.
Or maybe Aaron Cherian in our group could try running commands.
But I don't know if either of those would result in a system we could
login to and investigate.  Too bad.

I guess leave this open for a while, see if it turns up again in a way
that we can login to.

My best theory is that there's something in some user's environ,
either CONFIG_SITE or similar that pushes the libiberty layout into a
strange state.

Probably, I could make an artificial CONFIG_SITE that would cause a
problem.  The question is to find a not unreasonable value that
triggers the problem.

**Comment by hainest** [comment]
I'm happy to try to debug this with someone, if they can get an environment I can access.

**Comment by joennlae** [comment]
I have the same problem. It builds without any problem on my `gcc-9.3.0; Ubuntu 18.04 x86_64` system.

But I have the same error as @mwkrentel has on my `(custom built) gcc-9.3.1; Ubuntu 18.04 aarch64` system

```bash
     142    -- Could NOT find LibIberty (missing: LibIberty_LIBRARIES)
  >> 143    CMake Error at cmake/LibIberty.cmake:53 (message):
     144      LibIberty not found and cannot be downloaded because build is sterile.
     145    Call Stack (most recent call first):
     146      CMakeLists.txt:34 (include)
     147    
     148    
     149    -- Configuring incomplete, errors occurred!
```

I also came to the conclusion that adding `LibIberty_LIBRARIES` is necessary.
I know that my (custom built) gcc is built without `libiberty.a`  and I assume @mwkrentel one's is too (but this is just a guess). I know that not many users will be affected by this change but I believe this change will add more build stability. 

I have a theory (but I could also be wrong). That this

> (5) I've checked other machines where it works and theta-gpu. Neither
of them have /usr/include/libiberty.h or /usr/lib64/libiberty.a, so I
don't think the problem is that it's actually picking up a system
libiberty and not telling anyone.

could be the problem. That normally everyone who has `gcc` installed has also `libiberty` installed and that is why it was never a problem to build without the `LibIberty_LIBRARIES` variable.

I added a pull request to `spack`:

https://github.com/spack/spack/pull/25779

Feel free to change or reject it.

**Comment by hainest** [comment]
@joennlae Where is your libiberty.a located?

**Comment by joennlae** [comment]
@hainest Sorry for the long delay. I only have access to the ARM machines at work and we had a migration over the weekend.

Here: `./opt/spack/linux-ubuntu18.04-aarch64/gcc-9.3.1/libiberty-2.33.1-nnjhs3yzsuowkixryzzwyicm4rqxsbi4/lib64/libiberty.a
`

**Comment by hainest** [comment]
@joennlae That's weird! I've never had a problem with the libiberty from spack. Could you capture and post the output of the cmake config using `-DCMAKE_FIND_DEBUG_MODE=ON`, but without the `-DLibIberty_LIBRARIES=...`?
